### PR TITLE
Make DefaultProblemDetailsFactory public

### LINKED ISSUE #53030: Add a AddDefaultProblemDetailsFactory method that allows registering the internal `DefaultProblemDetailsFactory` class with the container
## Background and Motivation

Quote from #49982:
> I aim to leverage a `ProblemDetailsFactory` implementation (say the `DefaultProblemDetailsFactory` class) without adding MVC services to a Minimal API project.
> 
> More context: I'm updating older code to support .NET 7-8 that uses the `ProblemDetailsFactory` to generate `ProblemDetails` instances. The workaround was to call  `AddMvcCore` then, which was kinda ok since there was only MVC before. However, now that we have Minimal APIs, adding MVC dependencies to a non-MVC project makes no sense.

On top of that, the proposed API will bring the two Problem Details spaces one tiny step closer and more unified without introducing any breaking changes.

## Proposed API

The proposed API is to add a new `AddDefaultProblemDetailsFactory ` extension method that registers the default `DefaultProblemDetailsFactory` class and update the `AddMvcCore` method to leverage that extension instead.

Since the `Microsoft.AspNetCore.Http` abstractions seem like a better place than MVC to consolidate Problem Details classes, I suggest adding the extension method there if possible. The suggested `AddDefaultProblemDetailsFactory` method would look like this:

```csharp
namespace Microsoft.Extensions.DependencyInjection;
public static class DefaultProblemDetailsFactoryExtensions
{
    public static IServiceCollection AddDefaultProblemDetailsFactory(this IServiceCollection services)
    {
        services.TryAddSingleton<ProblemDetailsFactory, DefaultProblemDetailsFactory>();
        // Optional:
         services.TryAddEnumerable(ServiceDescriptor.Singleton<IProblemDetailsWriter, DefaultApiProblemDetailsWriter>());
    }
}
```

Update to MVC `AddMvcCoreServices` method:

```diff
// ...
namespace Microsoft.Extensions.DependencyInjection;
public static class MvcCoreServiceCollectionExtensions
{
    // ...
    internal static void AddMvcCoreServices(IServiceCollection services)
    {
        // ...
        // ProblemDetails
+       services.AddDefaultProblemDetailsFactory();
-       services.TryAddSingleton<ProblemDetailsFactory, DefaultProblemDetailsFactory>();
        // Optional: 
-       services.TryAddEnumerable(ServiceDescriptor.Singleton<IProblemDetailsWriter, DefaultApiProblemDetailsWriter>());
    }
}
```

See the [MvcCoreServiceCollectionExtensions](https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L276) class.

In case the team wants to consolidate the optional dependency from the preceding code blocks, they could also consider adapting the [`AddProblemDetails`](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/src/ProblemDetailsServiceCollectionExtensions.cs#L42) method of the `ProblemDetailsServiceCollectionExtensions` class to use this extension as well:

```diff
// ...
namespace Microsoft.Extensions.DependencyInjection;
public static class ProblemDetailsServiceCollectionExtensions
{
    // ...
    public static IServiceCollection AddProblemDetails(
        this IServiceCollection services,
        Action<ProblemDetailsOptions>? configure)
    {
        // ...
+       services.AddDefaultProblemDetailsFactory();
-        services.TryAddEnumerable(ServiceDescriptor.Singleton<IProblemDetailsWriter, DefaultProblemDetailsWriter>());
        // ...
    }
}
```

Of course, doing what the preceding code block suggests would add an additional dependency not used by the `ProblemDetailsService` class (the `ProblemDetailsFactory`) but would centralize the registering of the default `IProblemDetailsWriter`, which would guarantee that both MVC and Minimal APIs are aligned in their default implementation.

## Usage Examples

One could register the factory like this (`Program.cs`):

```csharp
builder.Services.AddDefaultProblemDetailsFactory();
```

## Alternative Designs

Please refer to the following issue for more info about alternate designs and thoughts put into this proposal in conjunction with @captainsafia: #49982.

## Risks

No foreseeable risks.

### PR DESCRIPTION
**Pull request by GiovanniBraconi**
- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

 Fixes #53030

Following this [suggestion](https://github.com/dotnet/aspnetcore/issues/53030#issuecomment-1877918398) i changed the DefaultProblemDetailsFactory class access modifier from internal to public.

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Infrastructure/DefaultProblemDetailsFactory.cs [hunk 1]
```diff
@@ -9,20 +9,37 @@
...
 
 namespace Microsoft.AspNetCore.Mvc.Infrastructure;
-
-internal sealed class DefaultProblemDetailsFactory : ProblemDetailsFactory
+/// <summary>
+/// The `DefaultProblemDetailsFactory` is a concrete implementation of the `ProblemDetailsFactory` abstract class.
+/// It provides methods to create instances of `ProblemDetails` and `ValidationProblemDetails` with default settings.
+/// This class uses the provided `ApiBehaviorOptions` for client error mapping and an optional custom configuration action to further customize the problem details.
+/// </summary>
+public sealed class DefaultProblemDetailsFactory : ProblemDetailsFactory
```
Comment (gfoidl) [inline src/Mvc/Mvc.Core/src/Infrastructure/DefaultProblemDetailsFactory.cs:17]: @GiovanniBraconi please address the analyzer-failure.
Have a look at [How to use Microsoft.CodeAnalysis.PublicApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) and see at the files in https://github.com/dotnet/aspnetcore/tree/main/src/Mvc/Mvc.Core/src folder.

File: src/Mvc/Mvc.Core/src/Infrastructure/DefaultProblemDetailsFactory.cs [hunk 2]
```diff
@@ -95,6 +108,6 @@ private void ApplyProblemDetailsDefaults(HttpContext httpContext, ProblemDetails
             problemDetails.Extensions["traceId"] = traceId;
         }
 
-        _configure?.Invoke(new() { HttpContext = httpContext!, ProblemDetails = problemDetails });
+        _configure?.Invoke(new() {HttpContext = httpContext!, ProblemDetails = problemDetails});
```
Comment (captainsafia) [inline src/Mvc/Mvc.Core/src/Infrastructure/DefaultProblemDetailsFactory.cs:111]: Nit: our style guidelines recommend spacing between braces here.

```suggestion
        _configure?.Invoke(new() { HttpContext = httpContext!, ProblemDetails = problemDetails });
```

**Review by captainsafia — APPROVED** [review]
LGTM outside of a small nit to apply.

We already have test coverage for this via [ProblemDetailsFactoryTest](https://source.dot.net/#Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ProblemDetailsFactoryTest.cs,e16b710dae9e3e36).

Thanks for the first look @gfoidl!

**Comment by uhfath** [comment]
Is there any chance of rolling this back to 8.x sometime in the future?

**Comment by martincostello** [comment]
New features aren't backported, so highly unlikely.

**Comment by uhfath** [comment]
It's a pity to hear that.

**Pull request merged**